### PR TITLE
[BEAM-2175] [BEAM-1115] Support for new State and Timer API in Spark batch mode

### PR DESCRIPTION
**Pull request by JingsongLi**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`.
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by JingsongLi** [comment]
R: @amitsela @aviemzur

**Comment by aviemzur** [comment]
Run Spark ValidatesRunner

**Comment by JingsongLi** [comment]
Run Spark ValidatesRunner

**Comment by aaltay** [comment]
Hey @JingsongLi any updates on this?

**Comment by JingsongLi** [comment]
Rebase it.

**Comment by JingsongLi** [comment]
retest this please

**Comment by amitsela** [comment]
R: @amitsela

**Comment by amitsela** [comment]
Run Spark ValidatesRunner

**Comment by amitsela** [comment]
retest this please

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/MultiDoFnFunction.java [hunk 1]
```diff
@@ -98,28 +110,108 @@ public MultiDoFnFunction(
 
     DoFnOutputManager outputManager = new DoFnOutputManager();
 
-    DoFnRunner<InputT, OutputT> doFnRunner =
+    SparkProcessContext.NoOpStepContext context;
```
Comment (amitsela) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/MultiDoFnFunction.java]: This should be of type `StepContext` and be `final`.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/MultiDoFnFunction.java [hunk 2]
```diff
@@ -98,28 +110,108 @@ public MultiDoFnFunction(
...
+    InMemoryTimerInternals timerInternals = null;
```
Comment (amitsela) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/MultiDoFnFunction.java]: `final`?

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/MultiDoFnFunction.java [hunk 3]
```diff
@@ -98,28 +110,108 @@ public MultiDoFnFunction(
...
+      Object key = null;
+      if (iter.hasNext()) {
+        WindowedValue<InputT> currentValue = iter.next();
+        key = ((KV) currentValue.getValue()).getKey();
+        iter = Iterators.concat(Iterators.singletonIterator(currentValue), iter);
+      }
+      final InMemoryStateInternals<?> stateInternals = InMemoryStateInternals.forKey(key);
+      timerInternals = new InMemoryTimerInternals();
+      final InMemoryTimerInternals finalTimerInternals = timerInternals;
+      context = new SparkProcessContext.NoOpStepContext(){
```
Comment (amitsela) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/MultiDoFnFunction.java]: This is not really a `NoOpStepContext` instance, but actually an implementation of `StepContext` (hence my comment about the type above).
```java
context = new StepContext(){
        @Override
        public StateInternals stateInternals() {
          return stateInternals;
        }

        @Override
        public TimerInternals timerInternals() {
          return timerInternals;
        }
      };
```
Might make more sense.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/MultiDoFnFunction.java [hunk 4]
```diff
@@ -98,28 +110,108 @@ public MultiDoFnFunction(
...
+    if (stateful) {
+      Object key = null;
+      if (iter.hasNext()) {
+        WindowedValue<InputT> currentValue = iter.next();
+        key = ((KV) currentValue.getValue()).getKey();
+        iter = Iterators.concat(Iterators.singletonIterator(currentValue), iter);
+      }
+      final InMemoryStateInternals<?> stateInternals = InMemoryStateInternals.forKey(key);
+      timerInternals = new InMemoryTimerInternals();
+      final InMemoryTimerInternals finalTimerInternals = timerInternals;
```
Comment (amitsela) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/MultiDoFnFunction.java]: The `timerInternals/finalTimerInternals` is a little confusing to me. How about having just the one - `final InMemoryTimerInternals timeInternals` and setting `null` accordingly? just as you set the `NoOpStepContext`.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/TransformTranslator.java [hunk 5]
```diff
@@ -347,41 +350,54 @@ public String toNativeString() {
...
+            doFn,
+            context.getRuntimeContext(),
+            transform.getMainOutputTag(),
+            transform.getAdditionalOutputTags().getAll(),
+            TranslationUtils.getSideInputs(transform.getSideInputs(), context),
+            windowingStrategy);
+
+        JavaPairRDD<TupleTag<?>, WindowedValue<?>> all;
+
+        DoFnSignature signature = DoFnSignatures.getSignature(transform.getFn().getClass());
```
Comment (amitsela) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/TransformTranslator.java]: How making the `isStateful` decision here and simply passing the boolean to the `MultiDoFnFunction` constructor? it does the exact same, right?

**Comment by JingsongLi** [comment]
Yes, I ran a simple WordCountByStatefulParDo on the YARN cluster.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/TransformTranslator.java (hunk 5 above)
Comment (JingsongLi) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/TransformTranslator.java]: Yes, it should be.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/MultiDoFnFunction.java (hunk 3 above)
Comment (JingsongLi) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/MultiDoFnFunction.java]: OK.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/MultiDoFnFunction.java (hunk 4 above)
Comment (JingsongLi) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/MultiDoFnFunction.java]: OK.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/MultiDoFnFunction.java (hunk 2 above)
Comment (JingsongLi) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/MultiDoFnFunction.java]: OK.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/MultiDoFnFunction.java (hunk 1 above)
Comment (JingsongLi) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/MultiDoFnFunction.java]: OK.

**Comment by JingsongLi** [comment]
retest this please

**Comment by amitsela** [comment]
Run Spark ValidatesRunner

**Comment by amitsela** [comment]
retest this please

**Comment by amitsela** [comment]
Run Spark ValidatesRunner

**Comment by amitsela** [comment]
LGTM, @aviemzur got any comments?

**Comment by aviemzur** [comment]
None from me, merging.

**Pull request closed without merging**